### PR TITLE
Use swatchdog to send message to gitter room

### PR DESCRIPTION
**Pull request by kencho51**
# Pull request for issue: #1132, #1133 and #1148

This is a pull request for the following functionalities:

1. Install the swatchdog tool.
2. Symbolic link the rclone_copy script log output to rclone_latest.log.
3. Make swatchdog capture the last line of migration log if it contains `ERROR`.
4. Pass that `ERROR` line to `send_notification.sh` which will then be pushed to the gitter room using gitter api.

### For testing on the dev environment
```
% cd gigadb/app/tools/wasabi-migration
% cp env.example .env
% docker-compose run --rm config
# Spin up the log monitoring service 
% docker-compose up -d swatchdog
# Check the swatchdog state
% docker-compose ps 
            Name                          Command               State   Ports
-----------------------------------------------------------------------------
wasabi-migration_swatchdog_1   swatchdog -c /app/config/s ...   Up  
# To generate log file containing ERROR
% docker-compose run --rm rclone /app/rclone_copy.sh --starting-doi 100001 --ending-doi 100320
# Check the log file can be found in the logs/ dir
# Check the ERROR message in the gitter room
# Stop and rm rclone container 
% docker stop $(docker ps -aq) && docker rm $(docker ps -aq)
# Or execute the wrapper script
% ./start_backup_process.sh 100001 100320 100
```

### For testing on the cngb backup server
Prerequisite:
1. Generate the Migration Wasabi sub-user configuration according to [here](https://github.com/pli888/gigadb-website/blob/rclone-cngb2wasabi/gigadb/app/tools/wasabi-migration/README.md#using-migration-tool-on-cngb-backup-server).
```
# Go to gigadb/app/tools/
% cd ..
% tar -czvf wasabi-migration.tar.gz wasabi-migration
% mv wasabi-migration.tar.gz wasabi-migration
# Go to gigadb/app/tools/wasabi-migration
% cd wasabi-migration
# Build the images customized for cngb backup server
% docker build -t rclone .
% docker tag rclone:latest rclone:cngb
% docker build -f Swatchdog-Dockerfile -t swatchdog .
% docker tag swatchdog:latest swatchdog:cngb
% docker images
REPOSITORY                       TAG                  IMAGE ID       CREATED         SIZE
rclone                           cngb                 ccd22e890601   7 minutes ago   61.3MB
rclone                           latest               ccd22e890601   7 minutes ago   61.3MB
swatchdog                        cngb                 44d987ded07f   12 days ago         60.3MB
swatchdog                        latest               44d987ded07f   12 days ago         60.3MB

% docker save rclone:cngb | gzip > rclone_cngb.tar.gz
% docker save swatchdog:cngb | gzip > swatchdog_cngb.tar.gz
% ls *.gz                  
rclone_cngb.tar.gz      swatchdog_cngb.tar.gz   wasabi-migration.tar.gz
# Copy *gz files to `/home/gigadb/ken` in cngb backup server through the smoc
```

2. Then log into the  cngb backup server
```
[gigadb@cngb-gigadb-bak ken]$ ls
README.md    rclone.conf             send_notification.sh   test-download-from-tyo  wasabi-migration.tar.gz
cronjob.txt  rclone_cngb.tar.gz      swatchdog              test-download-from-us1
logs         rclone_ken_cngb.tar.gz  test-download-from-sg  test_scripts
[gigadb@cngb-gigadb-bak ken]$ docker images
REPOSITORY                       TAG                 IMAGE ID            CREATED             SIZE
alpine                           3.16                bfe296a52501        4 weeks ago         5.54 MB
hello-world                      latest              feb5d9fea6a5        14 months ago       13.3 kB
rija/docker-alpine-shell-tools   1.0.1               3bdd20ea7a29        4 years ago         15.3 MB
[gigadb@cngb-gigadb-bak ken]$ docker load < rclone_cngb.tar.gz 
44da9b6c67d9: Loading layer 56.09 MB/56.09 MB
[gigadb@cngb-gigadb-bak ken]$ docker load < swatchdog_cngb.tar.gz 
Loaded image: swatchdog:cngb
[gigadb@cngb-gigadb-bak ken]$ docker images
REPOSITORY                       TAG                 IMAGE ID            CREATED             SIZE
rclone                           cngb                ccd22e890601        20 minutes ago      61.3 MB
swatchdog                        cngb                44d987ded07f        12 days ago         60.3 MB
alpine                           3.16                bfe296a52501        5 weeks ago         5.54 MB
hello-world                      latest              feb5d9fea6a5        15 months ago       13.3 kB
rija/docker-alpine-shell-tools   1.0.1               3bdd20ea7a29        4 years ago         15.3 MB
[gigadb@cngb-gigadb-bak ken]$ 
[gigadb@cngb-gigadb-bak ken]$ tar -xvf wasabi-migration.tar.gz
[gigadb@cngb-gigadb-bak ken]$ cd wasabi-migration
[gigadb@cngb-gigadb-bak wasabi-migration]$ ls
Dockerfile  config              env.example         logs               rclone.conf.dist  send_notification.sh  swatchdog.conf.dist
README.md   docker-compose.yml  generate_config.sh  proxy_settings.sh  rclone_copy.sh    spin_up_swatchdog.sh  tests
[gigadb@cngb-gigadb-bak wasabi-migration]$ docker-compose run --rm rclone_cngb rclone version
rclone v1.60.0
- os/version: alpine 3.16.3 (64 bit)
- os/kernel: 3.10.0-862.14.4.el7.x86_64 (x86_64)
- os/type: linux
- os/arch: amd64
- go/version: go1.19.2
- go/linking: static
- go/tags: none
[gigadb@cngb-gigadb-bak wasabi-migration]$ 
[gigadb@cngb-gigadb-bak wasabi-migration]$ docker-compose run --rm swatchdog_cngb swatchdog --version
This is swatchdog version 3.2.4
Built on Aug 25, 2008
Built by E. Todd Atkins <Todd.Atkins@StanfordAlumni.ORG>
[gigadb@cngb-gigadb-bak wasabi-migration]$ docker-compose run --rm rclone_cngb bash -c 'source /app/proxy_settings.sh; rclone -vv ls wasabi:gigadb-datasets/dev'
2022/12/07 15:38:00 DEBUG : rclone: Version "v1.60.0" starting with parameters ["rclone" "-vv" "ls" "wasabi:gigadb-datasets/dev"]
2022/12/07 15:38:00 DEBUG : Creating backend with remote "wasabi:gigadb-datasets/dev"
2022/12/07 15:38:00 DEBUG : Using config file from "/root/.config/rclone/rclone.conf"
    25964 dermatology.data
    39975 pub/10.5524/100001_101000/100002/CR.kegg.gz
      649 pub/10.5524/100001_101000/100002/readme.txt
     2696 pub/10.5524/100001_101000/100012/Keller.gff
     3164 pub/10.5524/100001_101000/100012/readme.txt
       34 test.txt
       57 tests/2022-11-18-1/sample.txt 
[gigadb@cngb-gigadb-bak wasabi-migration]$ 
# Spin up the swatchdog to watch for the rclone_latest.log
[gigadb@cngb-gigadb-bak wasabi-migration]$ docker-compose up -d swatchdog_cngb
Creating wasabi-migration_swatchdog_cngb_1 ... done
[gigadb@cngb-gigadb-bak wasabi-migration]$ docker-compose ps
              Name                             Command               State   Ports
----------------------------------------------------------------------------------
wasabi-migration_swatchdog_cngb_1   swatchdog -c /app/config/s ...   Up 
# Create log contains ERROR
[gigadb@cngb-gigadb-bak wasabi-migration]$ docker-compose run --rm rclone_cngb /app/rclone_copy.sh --starting-doi 100002 --ending-doi 100320
# Make sure the log 
[gigadb@cngb-gigadb-bak wasabi-migration]$ ls logs/
migration_20221212_134526.log  rclone_latest.log
# Check the error notification in gitter chat room, which should be started with `From server cngb-gigadb-bak : `
# Stop and rm rclone container 
[gigadb@cngb-gigadb-bak wasabi-migration]$ docker stop $(docker ps -aq) && docker rm $(docker ps -aq)
# Or execute the wrapper script
[gigadb@cngb-gigadb-bak wasabi-migration]$ ./start_backup_process.sh 100001 100320 100

```

## How have functionalities been implemented?

1. Every time executing the `rclone_copy.sh`, a log in a file name format `migration_yyyymmdd_hhmmss.log` will be created and be symbolic linked to `rclone_latest.log`.
3. swatchdog will watch the`rclone_latest.log` which points to the latest migration log.
4. swatchdog default behaviour `tail -n 0 -f`  is to capture the last line of the latest migration log and also look for newly added lines after the swatchdog start.
5. If the newly added line contains string pattern ` ERROR  :`, it will be captured by the swatchdog, then it will be passed to the `send_notification` script.
6. Separate the swatchdog service from the rclone container, which would make the swatchdog more versatile as it can keep running in the background like `docker-compose up -d swatchdog` and always looks for the latest changes in the `rclone_latest.log`. 
7. The `send_notification` script makes use of gitter api to push the error message to the designated gitter chat room.
8. A wrapper script `start_backup_process.sh` is created to chain all the steps, it would create new swatchdog service for every backup process, and remove it if the backup process executed successfully.

## Any issues with implementation?

1. `GITTER_API_TOKEN` and `GITTER_IT_NOTIFICATION_ROOM_ID` are found in the `FORK_VARIABLES`, where upstream user `Migration` will not get it. So these two variables have been relocated to `MISC_VARIABLES`, so they are now available for the upstream user.
2. swatchdog `--daemon` cannot be spinned up in a container, so a continuous changes in a particular log in background can be achieved by using `docker-compose up -d swatchdog`. 
3. When `rclone_copy.sh` started again, a newer (current) migration log would be produced, it has be to be linked to the `rclone_latest.log` as the swatchdog only watches for it.

## Any changes to automated tests?
None.

## Any changes to documentation?

Update swatchdog usage in readme.

## Any technical debt repayment?

None.

## Any improvements to CI/CD pipeline?

None.

### CODE REVIEW

File: gigadb/app/tools/wasabi-migration/Dockerfile [hunk 1]
```diff
@@ -10,6 +10,7 @@ RUN apk update \
         ca-certificates \
         fuse \
         curl \
+        swatch \
```
Comment (pli888) [inline gigadb/app/tools/wasabi-migration/Dockerfile]: I would prefer that the Swatchdog Docker service be created by a simpler `Swatchdog-Dockerfile` which just contains the `swatch` tool installed and has a `CMD` attribute which returns the version of Swatchdog installed.

File: gigadb/app/tools/wasabi-migration/docker-compose.yml [hunk 2]
```diff
@@ -15,10 +15,14 @@ services:
     volumes:
       - ./config:/root/.config/rclone
       - .:/app
-        
+    command: /app/spin_up_swatchdog.sh
```
Comment (pli888) [inline gigadb/app/tools/wasabi-migration/docker-compose.yml]: It feels confusing to use the `rclone` service to create the Swatchdog container and start the tool. I think it's better to have a specific Swatchdog Docker service called something like `rclone_monitor` defined in `docker-compose.yml`. You could even think about starting the Swatchdog service in `docker-compose.yml` using something similar to below:
```
command: swatchdog -c /app/config/swatchdog.conf -r $RESTART_TIME -t "$$(cat /app/logs/latest_rclone_log)"
```

File: gigadb/app/tools/wasabi-migration/README.md [hunk 3]
```diff
@@ -295,4 +328,20 @@ This can be overridden using the `--max-batch-size`. For example, to increase
...
+# To generate log file containing ERROR
+[gigadb@cngb-gigadb-bak wasabi-migration]$ docker-compose run --rm rclone_cngb /app/rclone_copy.sh --starting-doi 100002 --ending-doi 100320 
+# Check the log file can be found in the logs/ dir
+# Spin up the swatchdog 
+[gigadb@cngb-gigadb-bak wasabi-migration]$ docker-compose up -d rclone_cngb 
+# Stop and remove rclone container 
+[gigadb@cngb-gigadb-bak wasabi-migration]$ docker stop $(docker ps -aq) && docker rm $(docker ps -aq)
+# Or chain the cmd together, so the swatchdog could recognize the latest log and capture it, and then send it gitter room
+[gigadb@cngb-gigadb-bak wasabi-migration]$ docker-compose run --rm rclone_cngb /app/rclone_copy.sh --starting-doi 100002 --ending-doi 100320 && docker-compose up -d rclone_cngb && docker stop $(docker ps -aq) && docker rm $(docker ps -aq)
+Creating wasabi-migration_rclone_cngb_1 ... done
```
Comment (pli888) [inline gigadb/app/tools/wasabi-migration/README.md]: I am able to manually execute these commands to see the functionality working. I wonder if these commands would be better placed in a bash script for execution? The bash script would contain logic to check if a Swatchdog container is already running and, if so, then could use `docker restart` to get it to monitor the latest rclone_copy log file. If not, then the script would spin up a new Swatchdog container. I also wonder if it is a good idea to use `&&` to chain the commands together. If `rclone_copy` is long running then Swatchdog would not run until `rclone_copy` process has finished which could take days? Probably best that the Swatchdog container starts monitoring as soon as the new log file is present.

**Review by rija — CHANGES_REQUESTED** [review]
Hi @kencho51,

After doing:
```
cp env.example .env
```
And filling in the appropriate information, I was able to proceed with the "How to test section"

**praise**: I was able to follow through and the outcome in both scenarios (using the distinct commands and then using the wrapper script) led to appropriate entry in the log and appropriate notification appearing in the Gitter channel

**issue**: the following line (which is used three times) is dangerous because it's trying to stop any running docker containers, even those that have nothing to do with the "wasabi-migration_swatchdog" service. Because you are not running other containers, don't assume it's the case for the other developers (or a future you)
```
docker stop "$(docker ps -aq)" && docker rm "$(docker ps -aq)"
```

It's better to act on the particular service you want to stop:
```
docker-compose stop swatchdog || true
```
(You can still use ``up -d`` after the service has been stopped that way to restart it)

In general, when possible, avoid mixing usage of ``docker`` and ``docker-compose``, because their scope of operation is different. The latter is limited to the directory/context of where the docker-compose file is and is application/service level. The former is global and lower-level.

**issue (non-blocking)**: With Bash ``if`` conditional statement, it is thrown upon to have the ``then`` on a separate line.

Check out the examples:

* in Bash's official documentation: https://www.gnu.org/software/bash/manual/html_node/Conditional-Constructs.html
* And in the Linux documentation project: https://tldp.org/HOWTO/Bash-Prog-Intro-HOWTO-6.html

**question**: When I run the wrapper I got this output:
```
$ ./start_backup_process.sh 100001 100320 100
Error response from daemon: 404 page not found
wasabi-migration_swatchdog_1 is up-to-date
Creating wasabi-migration_rclone_run ... done
Error response from daemon: 404 page not found
```

I don't know what those "404 page not found" errors are doing here?
I think it is related to the dangerous line mentioned above.

**question**: in the final housekeeping statement,  the dangerous line can be replaced with ``docker-compose rm -fsv wasabi-migration_watchdog``
But I'm not sure the value of removing the service, since we may need to start the process multiple time?

**Comment by kencho51** [comment]
Hi @rija,

> But I'm not sure the value of removing the service, since we may need to start the process multiple time?

In the latest commits, the swatchdog service will be stopped instead of removing because it is meant to be started several times. If the variables `devBackupStatus` , `liveBackupStatus` are not empty and equals to `0`, the swatchdog service will be stopped. 

And the error `Error response from daemon: 404 page not found` has been fixed after placing the `then` at the same line as the if clause.

**Review by rija — APPROVED** [review]
Hi @kencho51, 
happy about the changes.
Providing that @pli888 has reviewed this and successfully tested it on CNGB infrastructure, I'm happy to approve.

File: gigadb/app/tools/wasabi-migration/start_backup_process.sh [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+  docker-compose run --rm rclone /app/rclone_copy.sh --starting-doi "$startingDoi" --ending-doi "$endingDoi" --max-batch-size "$maxBatchSize"
+  devBackupStatus=$?
+fi
+
+liveBackupStatus=""
+if [[ $serverName == "cngb-gigadb-bak" ]]; then
+  if [[ $(docker-compose ps | grep "wasabi-migration_swatchdog_cngb" | grep -c "Up") -gt 0 ]]; then
+      # stop existing swatchdog service, if any
+      docker-compose stop swatchdog || true
+    fi
```
Comment (pli888) [inline gigadb/app/tools/wasabi-migration/start_backup_process.sh]: Does the indentation need fixing in lines 31-33?

**Review by pli888 — APPROVED** [review]
I can get the functionality working on the CNGB backup server and the code looks fine to me - just take a look at my comment re: indentation and this PR can be merged.

**Pull request merged**